R/install.R

```diff
@@ -1,276 +1,343 @@
+  ## TODO: can we select a binary that is depended on by a source package?
 
-  # Remove installed dependencies from the dependency lists.
-  plan$dependencies <- lapply(plan$dependencies, setdiff, plan$package[installed])
+  ## Otherwise select a binary if there is one
+  can_install <- which(
+    state$plan$build_done &
+    ! state$plan$install_done &
+    is.na(state$plan$worker_id))
 
-  # Return uninstalled packages
-  plan[!installed, ]
-}
+  if (any(can_install)) {
+    pkgidx <- can_install[1]
+    return(task("install", pkgidx = pkgidx))
+  }
 
-get_output <- function(x) {
-  res <- poll(x$processes, -1)
-  idx <- which(vapply(res, function(x) any(x == "ready"), logical(1)))
-  lapply(idx,
-    function(i) {
-      list(
-        output = remove_spaces(x$processes[[i]]$read_output_lines()),
-        error = x$processes[[i]]$read_error_lines(),
-        index = i)
-    })
+  ## Looks like nothing else to do
+  task("idle")
 }
 
-running_processes <- function(x) {
-  length(x$processes) > 0
+task <- function(name, ...) {
+  list(name = name, args = list(...))
 }
 
-kill_all_processes <- function(events) {
-  for (proc in events$processes) {
-    proc$kill(tools::SIGINT)
+start_task <- function(state, task) {
+  if (task$name == "idle") {
+    state
+
+  } else if (task$name == "build") {
+    start_task_build(state, task)
+
+  } else if (task$name == "install") {
+    start_task_install(state, task)
+
+  } else {
+    stop("Unknown task, internal error")
   }
 }
 
-remove_spaces <- function(x) {
-  x <- x[!grepl("^[[:space:]]*$", x)]
-}
+get_worker_id <- (function() {
+  id <- 0
+  function() {
+    id <<- id + 1
+    as.character(id)
+  }
+})()
 
-#' @importFrom crayon make_style strip_style
-greyish <- function() make_style("darkgrey")
+#' @importFrom pkgbuild pkgbuild_process
 
-green_tick <- function() green(symbol$tick)
-red_cross <- function() red(symbol$cross)
+start_task_build <- function(state, task,
+                             dummy = FALSE, dummy_args = list()) {
+  pkgidx <- task$args$pkgidx
+  path <- state$plan$file[pkgidx]
+  vignettes <- state$plan$vignettes[pkgidx]
+  needs_compilation <- !identical(state$plan$needs_compilation[pkgidx], "no")
+  tmp_dir <- create_temp_dir()
+  lib <- state$config$lib
 
-#' @importFrom crayon cyan reset green
-#' @importFrom prettyunits pretty_dt
-#' @importFrom clisymbols symbol
-#' @export
-format.pkginstall_installed <- function(x, ...) {
-  glue("{green_tick()} Installed {greyish()}{x$package} {cyan}({pretty_dt(x$time)}){reset}")
-}
+  px <- if (dummy) {
+    do.call(make_dummy_worker_process, dummy_args)
 
-#' @export
-format.pkginstall_built <- function(x, ...) {
-  glue("{green_tick()} Built {greyish()}{x$package} {cyan}({pretty_dt(x$time)}){reset}")
+  } else {
+    pkgbuild_process$new(
+      path, tmp_dir, binary = TRUE, vignettes = vignettes,
+      needs_compilation = needs_compilation, compile_attributes = FALSE,
+      args = glue("--library={lib}"))
```

Comment:
Do we need to worry about quoting the library directory if there are spaces in the path?

Reply:
No, that's what's great in callr/processx.